nodeup/pkg/model/docker_test.go

```diff
@@ -86,34 +88,80 @@ func TestDockerPackageHashes(t *testing.T) {
 
 	for _, dockerVersion := range dockerVersions {
 		t.Run(dockerVersion.Source, func(t *testing.T) {
-			verifyPackageHash(t, dockerVersion.Source, dockerVersion.Hash)
+			if err := verifyPackageHash(dockerVersion.Source, dockerVersion.Hash, dockerVersion.Version); err != nil {
+				t.Errorf("error verifying package %q: %v", dockerVersion.Source, err)
+			}
 
 			for _, p := range dockerVersion.ExtraPackages {
-				verifyPackageHash(t, p.Source, p.Hash)
+				if err := verifyPackageHash(p.Source, p.Hash, p.Version); err != nil {
+					t.Errorf("error verifying package %q: %v", p.Source, err)
+				}
 			}
 		})
 	}
 }
 
-func verifyPackageHash(t *testing.T, u string, hash string) {
-	resp, err := http.Get(u)
+func verifyPackageHash(u string, hash string, expectedVersion string) error {
+	name := path.Base(u)
+	p := filepath.Join("/tmp", name)
+
+	expectedHash, err := hashing.FromString(hash)
 	if err != nil {
-		t.Errorf("%s: error fetching: %v", u, err)
-		return
+		return err
 	}
-	defer resp.Body.Close()
 
-	hasher := sha1.New()
-	if _, err := io.Copy(hasher, resp.Body); err != nil {
-		t.Errorf("%s: error reading: %v", u, err)
-		return
+	if _, err := fi.DownloadURL(u, p, expectedHash); err != nil {
+		return err
 	}
 
-	actualHash := hex.EncodeToString(hasher.Sum(nil))
-	if hash != actualHash {
-		t.Errorf("%s: hash was %q", u, actualHash)
-		return
+	actualHash, err := hashing.HashAlgorithmSHA1.HashFile(p)
+	if err != nil {
+		return fmt.Errorf("error hashing file: %v", err)
 	}
+
+	if hash != actualHash.Hex() {
+		return fmt.Errorf("hash was %q, expected %q", actualHash.Hex(), hash)
+	}
+
+	if strings.HasSuffix(u, ".deb") {
+		cmd := exec.Command("dpkg-deb", "-I", p)
+		out, err := cmd.CombinedOutput()
+		if err != nil {
+			return fmt.Errorf("error running 'dpkg-deb -I %s': %v", p, err)
+		}
+
+		version := ""
+		for _, line := range strings.Split(string(out), "\n") {
+			line = strings.TrimSpace(line)
+			if !strings.HasPrefix(line, "Version: ") {
+				continue
+			}
+			version += strings.TrimPrefix(line, "Version: ")
+		}
+		if version != expectedVersion {
+			return fmt.Errorf("unexpected version, actual=%q, expected=%q", version, expectedVersion)
+		}
+
+	} else if strings.HasSuffix(u, ".rpm") {
+		cmd := exec.Command("rpm", "-qp", "--queryformat", "%{VERSION}", "--nosignature", p)
+		out, err := cmd.CombinedOutput()
+		if err != nil {
+			return fmt.Errorf("error running rpm %s: %v", strings.Join(cmd.Args, " "), err)
+		}
+
+		version := strings.TrimSpace(string(out))
+		if version != expectedVersion {
+			return fmt.Errorf("unexpected version, actual=%q, expected=%q", version, expectedVersion)
+		}
+	} else if strings.HasSuffix(u, ".tgz") {
```

Comment:
Packages for conatinerd have a 'tar.gz' suffix:
https://github.com/kubernetes/kops/pull/8199/files#diff-ecf61990c0cdda1c13ad07df0440a537R116-R135
```suggestion
	} else if strings.HasSuffix(u, ".tgz") || strings.HasSuffix(u, ".tar.gz") {
```

Reply:
Good call - added!